### ISSUE TITLE
LoRRC: Low-rank Recurrent Convolutional Random Number Generator

### ISSUE THREAD

**Issue by AStarCale** [AStarStartup/LinearId#71]
# 

## Problem

The problem is that it takes a lot of CPU power and time to generate uniform random numbers..

## Solution

The solution invented by Cale McCollough is we store use a CPU based neural network with State-machine based math engine that does e^x and square root calculations each time I generate  UUID storing a running XOR, and I store the last random client-generated pseudorandom or quasirandom or local pseudorandom generated random numbers -1 < x < 1 in index 0 through 11, hereinafter dirty inputs. Index 12 through 23 contain the last Output A, and 24 through 35 contain the last Output B in ring buffers  To generate a random number I first XOR the mantissa of the running XOR with the the last client-generated or pseudo random number (exponent must be< 0) and add that to the last 11 dirty inputs then divide by 12. If this number forms a normal distribution distribution (i.e. bell curve), then this is the outputted random number.

Else I perform a 3x3 determinant sliding window convolution filter to reduce the 36 inputs to 10 neural network inputs, and I use an NxN LoRA split in half diagonally such that each half forms a neural network that reduces by one dimension each layer and has one output. Half is used to generate the Output A, the other half used to generate output B.

The system keeps track of three sets of NxN neural network values, Checkpoint A, Checkpoint B, and Weights-Bias Max. Stochastic Gradient Descent is achieved using a set of LoRA velocity vectors. Pseudorandom bits are generated by XORing X of the neural network inputs together, possibly looping around the entire memory map 12 at a time, or maybe only through the weights and biases. I use four pseudorandom bits to determine if the LoRA velocity will 25% increase, 25% decrease, or 50% not move. I store these bits in History 1 through History N. If the current path reduces in probability, I can revert back to random places in the history.

I cycle back and forth between Checkpoint A and Checkpoint B infinitely.

We use sample values -1 < x < 1 because this is a 63-bit random number. When you add the 12 dirty inputs and divide by 24 if you add signed probabilities, and it's 62-bit if the probabilities are unsigned and less then 1, but in  IEEE 754 standard value for 1 is the integer value 1, and if the exponent is -1 and the mantissa is all ones, that value is almost one but not quiet. If we can added up random samples between the max and min negative float values, we would require a 128-bit floating point number for intermediate calculations for the benefit of only one random bit. There are a lot of ways to get one random bit in a computer, so it's much more processor efficient to bound -1 to 1, and we just steal that .

To train the neural network, the algorithm expects an output of -1, 1, and 0 +/- the smallest float delta with exponent making the hidden bit the first fractional bit, alternating each side of the Split LoRA is used each time.. If the output was less than -1 the target output is set to -0-delta. If the output was aver 1 then the target is set to 0 + delta. If the output doesn't match a normal distribution using Polar rejection, another pseudo random value, a pseudo random value that does pass polar rejection is used.

## LoRRC Memory Map

```AsciiArt
# LoRRC Memory Map
+----------------------+ 
|     Dirty Inputs     | 0x0 
|     Clean Inputs     |  |
|       Output A       |  +
|       Output B       |
|  Optional Output C   |  <-- For Quad LoRA Split
|  Optional Output D   |  <-- For Quad LoRA Split
|    LoRA Velocity     |
|    LoRA History 0    |
|    LoRA History 1    |
|         ...          |
|    LoRA History N    |
| SoftMax Calculations |
|  Layer Calculations  |
|     Checkpoint A     |
|     Checkpoint B     |
|    Checkpoint Max    |
+----------------------+
```

### File Affected

1. `**/*.*`

## Hierarchy

* #10

## Tags

Mission

### Sessions

* AStarCale/AStarCale#9

## License

Copyright [AStarship](https://astarhip.net); all rights reserved.
